urllib3/fields.py

```diff
@@ -31,45 +31,93 @@ def format_header_param(name, value):
         The name of the parameter, a string expected to be ASCII only.
     :param value:
         The value of the parameter, provided as a unicode string.
+    :ret:
+        An RFC-2231-formatted unicode string.
     """
     if not any(ch in value for ch in '"\\\r\n'):
-        result = '%s="%s"' % (name, value)
+        result = u'%s="%s"' % (name, value)
         try:
             result.encode('ascii')
         except (UnicodeEncodeError, UnicodeDecodeError):
             pass
         else:
             return result
-    if not six.PY3 and isinstance(value, six.text_type):  # Python 2:
+
+    if not six.PY3:  # Python 2:
         value = value.encode('utf-8')
+
+    # encode_rfc2231 accepts an encoded string and returns an ascii-encoded
+    # string in Python 2 but accepts and returns unicode strings in Python 3
     value = email.utils.encode_rfc2231(value, 'utf-8')
     value = '%s*=%s' % (name, value)
+
+    if not six.PY3:  # Python 2:
+        value = value.decode('utf-8')
+
     return value
 
 
+def format_header_param_html5(name, value):
+    """
+    Helper function to format and quote a single header parameter.
+
+    Particularly useful for header parameters which might contain
+    non-ASCII values, like file names. This follows the HTML5 Working Draft
+    Section 4.10.22.7 (http://w3c.github.io/html/sec-forms.html#multipart-form-data)
+
+    :param name:
+        The name of the parameter, a string expected to be ASCII only.
+    :param value:
+        The value of the parameter, provided as a unicode string.
+    :ret:
+        A unicode string, stripped of troublesome characters.
+    """
+
+    value = value.replace(u'\\', u'\\\\').replace(u'"', u'\\"')
+    value = value.replace(u'\r', u' ').replace(u'\n', u' ')
+
+    return u'%s="%s"' % (name, value)
+
+
+format_header_param = format_header_param_rfc2231  # For backwards-compatibility
+
+
 class RequestField(object):
+    DEFAULT_FILENAME_ENCODING_STYLE = 'HTML5'
```

Comment:
That is, something like:

```python
class RequestField(object):
    header_formatter = format_header_param_html5
```

I don't think this is the kind of thing that will be changing from request to request, so it's probably safe avoiding passing it in the constructors? Just let people override the static configuration-style variable in the class if they want to change the behaviour.